File: Baekjoon/java/Algorithm/src/baekjoon/BJ_토마토_7576_S1.java
package baekjoon;

import java.io.*;
import java.util.*;

public class BJ_토마토_7576_S1 {
	static int N,M,cnt,result;
	static int[][] array;
	static boolean[][] visited;
	static int[] dx = {-1,0,1,0};
	static int[] dy = {0,1,0,-1};
	static boolean check;
	
	public static class Point {
		int x,y;
		
		public Point(int x, int y) {
			this.x = x;
			this.y = y;
		}
	}
	
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		N = Integer.parseInt(st.nextToken());
		M = Integer.parseInt(st.nextToken());
		
		array = new int[M][N];
		for (int i = 0; i < M; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				array[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		bfs();
		
		for (int i = 0; i < M; i++) {
			for (int j = 0; j < N; j++) {
				if(array[i][j] == 0) {
					check = true;
				}
				result = Math.max(result, array[i][j]);
			}
		}
		if(check) System.out.println("-1");
		else System.out.println(result-1);
	}

	private static void bfs() {
		Queue<Point> q = new LinkedList<Point>();
		
		for (int i = 0; i < M; i++) {
			for (int j = 0; j < N; j++) {
				if(array[i][j] == 1) {
					q.add(new Point(i,j));
				}
			}
		}
		
		while(!q.isEmpty()) {
			Point p = q.poll();
			for (int i = 0; i < 4; i++) {
				int nx = p.x+dx[i];
				int ny = p.y+dy[i];
				
				if(0<=nx && nx<M && 0<=ny && ny<N) {
					if(array[nx][ny] == 0) {
						array[nx][ny] = array[p.x][p.y]+1;
						q.add(new Point(nx,ny));
					}
				}
			}
		}
	}
}

File: Baekjoon/python/BJ_토마토_7576_S1.py
import sys
from collections import deque
input = sys.stdin.readline

M,N = map(int,input().split())
visited = [[False] * M for _ in range(N)]
dx = [-1,0,1,0]
dy = [0,1,0,-1]
array,q =[],deque()

for i in range(N):
    array.append(list(map(int,input().split())))
    for j in range(M):
        if(array[i][j] == 1):
            q.append((i,j))

def bfs():
    while q:
        x,y = q.popleft()

        for i in range(4):
            nx,ny = x+dx[i],y+dy[i]

            if 0<=nx<N and 0<=ny<M:
                if not visited[nx][ny] and array[nx][ny] == 0:
                    visited[nx][ny] = True
                    array[nx][ny] = array[x][y] + 1
                    q.append((nx,ny))

bfs()
result = 0
check = False
for i in range(N):
    if 0 in array[i]:
        check = True
        break
    result = max(max(array[i]),result)

if check:
    print(-1)
else:
    print(result-1)